google/cloud/storage/internal/async_connection_impl.h

```diff
@@ -0,0 +1,88 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_INTERNAL_ASYNC_CONNECTION_IMPL_H
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_INTERNAL_ASYNC_CONNECTION_IMPL_H
+
+#include "google/cloud/storage/idempotency_policy.h"
+#include "google/cloud/storage/internal/async_connection.h"
+#include "google/cloud/storage/internal/storage_stub.h"
+#include "google/cloud/storage/options.h"
+#include "google/cloud/storage/retry_policy.h"
+#include "google/cloud/completion_queue.h"
+#include "google/cloud/future.h"
+#include "google/cloud/version.h"
+#include <memory>
+
+namespace google {
+namespace cloud {
+namespace storage_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+class AsyncConnectionImpl : public AsyncConnection {
+ public:
+  explicit AsyncConnectionImpl(CompletionQueue cq,
+                               std::shared_ptr<StorageStub> stub,
+                               Options options);
+  ~AsyncConnectionImpl() override = default;
+
+  Options options() const override { return options_; }
+
+  future<Status> AsyncDeleteObject(
+      storage::internal::DeleteObjectRequest const& request) override;
+
+ private:
+  std::unique_ptr<storage::RetryPolicy> retry_policy() {
+    auto const& options = internal::CurrentOptions();
+    if (options.has<storage::RetryPolicyOption>()) {
```

Comment:
ugh, forgot to add this comment to my review.

The current options vs connection options behavior isn't tested. I decided to test it in bigtable, but [it was annoying](https://github.com/googleapis/google-cloud-cpp/blob/9c6256e77b03e83cee73faea2cf4880f807612df/google/cloud/bigtable/internal/data_connection_impl_test.cc#L113-L182). Your call on whether or not to add a test.